bitwizeshift/result
# Improve CMake usage

### Checklist
- [x] The coding style is consistent with the rest of the library
- [x] My branch's history is clean and coherent. This could be done through
      at least one of the following practices:
  * Rebasing my branch off of the branch being merged to
  * Squashing commits to create a more cohesive history
- [x] If relevant, I have included unit-tests (for new code/bugfixes)

----------

### Description

I realize that my commits don't live up to your standards and that's fine for me, there is only so much grunt work I'm willing to do and I haven't adapted the CI files yet.

Some things to note about the changes:
* The documentation falsely claimed that modern CMake starts with 3.4, however 3.4 hasn't added anything significant that would make it modern. Modern CMake starts with 3.8, as that is the version which added meta compile features to specify the C++ standard requirement as a lower bound. The `(CMAKE_)?\<LANG>_STANDARD` variables/properties define an upper bound, which may decay to earlier standards.
* The root project CML and test project CML files now only deal with their respective requirements. The root project CML must be nothing more than (almost) pure build and usage requirements of the project. Things like hardcoded conan details, compiler flags and alike are **not** requirements of the project, but the developer developing the library, and as such there is a separate superbuild CML file (in the `all` directory) that wires things together in a way that things work.
* I haven't actually found a way to cleanly represent subprojects' requirements from a root conanfile.py and I don't write conan recipes myself, so I have no clue how to have a root conanfile.py without the test subprojects' details poisoning it. For this reason, I just dropped a conanfile.txt for catch in the `test` directory and called it a day. This point is also similar to the above, as the conanfile.py describes your library, not your tests and the test_package tests whether your library generates the conan package as it should and can be used as it should, nothing about your unit tests, etc.

The following commands show how to run things, but you can also look at https://github.com/erikzenker/hsm/blob/master/.github/workflows/windows-ci.yml (this uses conan for its Boost dependency) and https://github.com/friendlyanon/cxx-static-shared-example/blob/master/.github/workflows/ci.yml (this uses a superbuild CML to build things in CI) for real world examples.

How to build and install the library with this PR:
```bash
cmake -S . -B build
# nothing to build, header only
cmake --install build
```

How to build the tests with this PR:
```bash
cmake -S test -B build
cmake --build build
(cd build; ctest)
```

How to do the above using the superbuild with this PR:
```bash
conan install test -if build/deps
cmake -S all -B build/all -D "CMAKE_MODULE_PATH=${PWD}/build/deps"
cmake --build build/all
cmake --install build/all
(cd build/all/sub/test; ctest)
```

How to do the above using the superbuild with this PR on Windows (or multi-config generators in general):
```batch
conan install test -if build\deps
cmake -S all -B build\all -D "CMAKE_MODULE_PATH=%cd:\=/%/build/deps"
cmake --build build\all --config Release
cmake --install build\all --config Release
%comspec% /C "cd build\all\sub\test & ctest -C Release"
```

Also, I found a bug with Catch2's conan package, because it does not provide the proper module path to enable `include(Catch)`. You should report this to them.

For rationale behind everything in the refactored CML files you can read around in my PR to https://github.com/Dobiasd/FunctionalPlus and my commit messages in https://github.com/friendlyanon/cxx-static-shared-example

So you may ask, how to move forward with this PR? Like I said, the list of requirements for PRs is gigantic and there is only so much mundane programming I'm willing to do. This PR shows a project structure to strive for that doesn't look like the build tooling was bolted on as an afterthough with no care put into it.  
You may take the code in this PR as is, split commits up, make them adhere to your personal guidelines.
